src/main/java/bio/terra/stairway/FlightDao.java

```diff
@@ -423,6 +434,66 @@ private void completeInner(FlightContext flightContext) throws SQLException {
     }
   }
 
+  /**
+   * Fatally fail the flight id.
+   *
+   * <p>This should only be used if there are problems creating the FlightContext such that {@link
+   * #complete(FlightContext)} cannot be used.
+   *
+   * <p>This does not do a hook state transition as we do not have a FlightContext.
+   *
+   * @param flightId flight to fail fatally.
+   * @param fatalCause the reason to fail the flight fatally. May be null.
+   */
+  void fatalFail(String flightId, Exception fatalCause)
+      throws DatabaseOperationException, InterruptedException {
+    String serializedException;
+    try {
+      serializedException = exceptionSerializer.serialize(fatalCause);
+    } catch (Exception e) {
+      // We really do not want to throw while we're trying to fatally fail the flight.
+      logger.error("Unable to serialize fatalFail exception.", e);
+      serializedException = "Cause unknown. Unable to serialize fatalFail exception.";
+    }
+    String finalSerializedException = serializedException;
```

Comment:
why do you need this reference to the string?

Reply:
Java lambdas, the string must be effectively final to be referenced in the retry lambda.

Reply:
ah